### DISCUSSION TITLE
How can I use the Solver with Java language?

### DISCUSSION THREAD

**Discussion by giuseppeboezio** [tuProlog/2p-kt#422] [Q&A] [answered]
I am trying to use the Solver with Java language, I have a custom Library called ClpFDLibrary and I want to activate the flag for tracking interesting variables. 
My code is the following:
```java
Solver solver = ClassicSolverFactory.solverOf(
            ClpFDLibrary.toRuntime(),
            FlagStore.EMPTY + TrackVariables { ON },
            theory
);
```
How can I fix the code in order to work with Java language?

**Comment by gciatto** [comment, marked as answer]
Quick answer:
```java
import it.unibo.tuprolog.solve.Solver;
import it.unibo.tuprolog.solve.flags.TrackVariables;
import it.unibo.tuprolog.solve.library.Runtime;
import package.of.MyLibrary; // assuming this is a Kotlin object

Solver solver = Solver.prolog().newBuilder()
        .runtime(Runtime.of(MyLibrary.INSTANCE))
        .flag(TrackVariables.INSTANCE, TrackVariables.ON)
        .staticKb(theory)
        .build();
```

An explanation will follow.

**Comment by gciatto** [comment]
Explanation:
1. `ClassicSolverFactory` is somewhat an implementation detail. The same factory can and should be achieved via `Solver.prolog()`, in Java (or simply `Solver.prolog` in Kotlin)

2. Since version `0.30.x`, the builder pattern can be exploited to build solvers. While this is somewhat equivalent to `Solver.factory().solver*Of(...)` methods in theory, in practice, builders are far handier in Java. The reason is that `SolverFactory`s rely on default values for their methods, which only exist in Kotlin, while Java does not support them.

3. So, the code I provided above is actually equivalent to the following Kotlin code:
    ```kotlin
    import it.unibo.tuprolog.solve.Solver
    import it.unibo.tuprolog.solve.flags.TrackVariables
    import it.unibo.tuprolog.solve.library.toRuntime
    import package.of.MyLibrary; // assuming this is a Kotlin object

    val solver: Solver = Solver.prolog.newBuilder()
        .runtime(MyLibrary.toRuntime())
        .flag(TrackVariables) { ON }
        .staticKb(theory)
        .build()
    ```
    but keeping into account that:

    1. Kotlin objects are compiled as ordinary Java classes with a public static field named `INSTANCE` and referencing the only admissible instance of that class
    2. the `Library.toRuntime()` method is actually an extension method, which cannot be invoked via infix notation in Java. In Java, one can attain the same result via `Runtime.of(library)`
    3. the `.flag(TrackVariables) { ON }` is not possible in Java ('cause it does not support the dangling lambda convention, so the best override is `.flag(library, value)`
    4. notice that notable flags such as `TrackVariables` are actually implemented as Kotlin objects too, so in Java they must be referenced as `TrackVariables.INSTANCE`
